lib/Dialect/HW/HWTypes.cpp

```diff
@@ -236,17 +238,21 @@ llvm::hash_code hash_value(const FieldInfo &fi) {
 } // namespace hw
 } // namespace circt
 
-/// Parse a list of field names and types within <>. E.g.:
+/// Parse a list of unique field names and types within <>. E.g.:
 /// <foo: i7, bar: i8>
 static ParseResult parseFields(AsmParser &p,
                                SmallVectorImpl<FieldInfo> &parameters) {
+  llvm::StringSet<> nameSet;
   return p.parseCommaSeparatedList(
       mlir::AsmParser::Delimiter::LessGreater, [&]() -> ParseResult {
         std::string name;
         Type type;
         if (p.parseKeywordOrString(&name) || p.parseColon() ||
             p.parseType(type))
           return failure();
+        if (!nameSet.insert(name).second)
+          return p.emitError(p.getCurrentLocation(),
+                             "duplicate field name \'" + name + "\'");
```

Comment:
Checking this during parsing and verification results means the check has to be duplicated. Is the error just as good if you ignore the parsing check and wait for the verifier to blow up?

Reply:
I originally just did the verify method. It does the trick, but a failure there will dump a full stack trace, which I found too verbose and somewhat intimidating. 😄

Reply:
Fair. One thing you can do to further improve this is to use a better location for the error. Currently the error looks like:

```
# circt-opt Foo.mlir 
Foo.mlir:2:65: error: duplicate field name 'foo'
    %0 = hw.struct_create (%a, %b) : !hw.struct<foo: i8, foo: i8, foo: i8>
                                                                ^
Foo.mlir:2:74: error: duplicate field name 'foo'
    %0 = hw.struct_create (%a, %b) : !hw.struct<foo: i8, foo: i8, foo: i8>
                                                                         ^
```

If you use:

```diff
diff --git a/lib/Dialect/HW/HWTypes.cpp b/lib/Dialect/HW/HWTypes.cpp
index e89d9b3ab..0c22baef2 100644
--- a/lib/Dialect/HW/HWTypes.cpp
+++ b/lib/Dialect/HW/HWTypes.cpp
@@ -248,13 +248,14 @@ static ParseResult parseFields(AsmParser &p,
       mlir::AsmParser::Delimiter::LessGreater, [&]() -> ParseResult {
         std::string name;
         Type type;
+
+        auto fieldLoc = p.getCurrentLocation();
         if (p.parseKeywordOrString(&name) || p.parseColon() ||
             p.parseType(type))
           return failure();
 
         if (!nameSet.insert(name).second) {
-          p.emitError(p.getCurrentLocation(),
-                      "duplicate field name \'" + name + "\'");
+          p.emitError(fieldLoc, "duplicate field name \'" + name + "\'");
           // Continue parsing to print all duplicates, but make sure to error
           // eventually
           hasDuplicateName = true;
```

You get:
```
# circt-opt Foo.mlir 
Foo.mlir:2:58: error: duplicate field name 'foo'
    %0 = hw.struct_create (%a, %b) : !hw.struct<foo: i8, foo: i8, foo: i8>
                                                         ^
Foo.mlir:2:67: error: duplicate field name 'foo'
    %0 = hw.struct_create (%a, %b) : !hw.struct<foo: i8, foo: i8, foo: i8>
                                                                  ^
```